### PR TITLE
[8163] WIP api reset cache with a helper query

### PR DESCRIPTION
**Pull request by m4ra**
**Describe your changes**
Briefly explain what you did and provide context for a clearer understanding.

**Tasks**
- [ ] PR name contains story or task reference
- [ ] Steps to recreate and test the changes
- [ ] Documentation (docs and inline)
- [ ] Tests (including n+1 and django_assert_num_queries where applicable)
- [ ] Changelog

### CODE REVIEW

File: meinberlin/apps/projects/api.py [hunk 1]
```diff
@@ -11,6 +11,28 @@
...
+        .select_related("administrative_district", "organisation")
+        .prefetch_related(
+            "moderators",
+            "plans",
+            "organisation__initiators",
+            "module_set__phase_set",
+        )
+    )
+    return projects
+
```
Comment (m4ra) [inline meinberlin/apps/projects/api.py:35]: I left this as the reference of the ORM filtering

File: meinberlin/apps/projects/api.py [hunk 2]
```diff
@@ -20,25 +42,7 @@ def __init__(self, *args, **kwargs):
...
-            .order_by("created")
-            .select_related("administrative_district", "organisation")
-            .prefetch_related(
-                "moderators",
-                "plans",
-                "organisation__initiators",
-                "module_set__phase_set",
-            )
-        )
+        projects = PublicProjectView.objects.all()
```
Comment (m4ra) [inline meinberlin/apps/projects/api.py:45]: this is now querying the virtual db table

File: meinberlin/apps/projects/helpers.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            a4modules_module module ON project.id = module.project_id
+        LEFT JOIN 
+            a4phases_phase phases ON module.id = phases.module_id
+        WHERE 
+            (project.project_type = 'a4projects.Project' OR project.project_type = 'meinberlin_bplan.Bplan')
+            AND (project.access = 1 OR project.access = 2)
+            AND project.is_draft = FALSE
+            AND project.is_archived = FALSE
+        ORDER BY 
+            project.created;"""
```
Comment (m4ra) [inline meinberlin/apps/projects/helpers.py:37]: I think I need a GROUP BY to add the start/end dates here. Have to look more into that.

File: meinberlin/apps/projects/helpers.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        LEFT JOIN 
+            a4projects_project_moderators moderator ON project.id = moderator.project_id
+        LEFT JOIN 
+            meinberlin_plans_plan plan ON project.id = moderator.project_id
+        LEFT JOIN 
+            meinberlin_organisations_organisation_initiators initiators ON organisation.id = initiators.organisation_id
+        LEFT JOIN 
+            a4modules_module module ON project.id = module.project_id
+        LEFT JOIN 
+            a4phases_phase phases ON module.id = phases.module_id
```
Comment (m4ra) [inline meinberlin/apps/projects/helpers.py:30]: Not sure if those are the correct query to get the phases, any ideas welcome. need to research more.

File: meinberlin/apps/projects/tasks.py [hunk 5]
```diff
@@ -191,5 +192,7 @@ def reset_cache_for_projects(starts: bool, ends: bool) -> str:
             msg += "failed for active projects becoming past"
         else:
             msg += "succeeded for active projects becoming past"
+    # update virtual table view
+    public_projects_exec_raw_query()
```
Comment (m4ra) [inline meinberlin/apps/projects/tasks.py:196]: when clearing the cash, the raw query it reconstructs the virtual table. But we could also make a separate task to do that few times in a day.

File: meinberlin/apps/projects/helpers.py (hunk 4 above)
Comment (goapunk) [reply meinberlin/apps/projects/helpers.py:30]: hmm, as this is already a rather complex query I'm not super convinced this is the easier way to go (i.e. having to maintain a raw sql query if we change anything on the models vs the ORM just giving us an error) compared to moving the query and the filtering into functions we can call from the cache task. What are your thoughts, I haven't worked a lot with raw sql in django before, so this seems way more complicated to me :/

**Comment by m4ra** [comment]
Closing this.

**Pull request closed without merging**